### PR TITLE
Binary graph fragments for BuildXL

### PR DESCRIPTION
**Pull request by narasamdya**
- Uses BuildXL script analyzer for creating graph fragments.
- Merge/unify pips, including IPC and service pips.

### CODE REVIEW

File: Public/Sdk/Public/Prelude/Prelude.IO.dsc [hunk 1]
```diff
@@ -48,6 +48,12 @@ declare function globFolders(folder: Directory, pattern?: string, isRecursive?:
 declare function globR(folder: Directory, pattern?: string): File[];
 declare function globRecursively(folder: Directory, pattern?: string): File[];
 
+/**
+ * Unsafe functions to create output file artifact and exclusive output directory.
+ */
+declare function unsafe_outputFile(path: Path, rewriteCount?: number): File;
+declare function unsafe_exOutputDirectory(path: Path): Directory;
+
```
Comment (dannyvv) [inline Public/Sdk/Public/Prelude/Prelude.IO.dsc]: Please place these in a namespace 'Unsafe' just so it is contained....

**Comment by dannyvv** [comment]
Can you split this in smaller chunks? You have several features in here that could be done as smaller commits i.e. Module

File: Public/Src/FrontEnd/Script/Ambients/AmbientContext.cs [hunk 2]
```diff
@@ -192,7 +192,7 @@ private EvaluationResult GetMount(Context context, ModuleLiteral env, Evaluation
             var name = Convert.ToString(args[0].Value, CultureInfo.InvariantCulture);
             GetProvenance(context, out AbsolutePath path, out LineInfo lineInfo);
 
-            var result = engine.TryGetMount(name, "Script", new BuildXL.Utilities.ModuleId(0), out IMount mount);
+            var result = engine.TryGetMount(name, "DominoScript", AmbientTypes.ScriptModuleId, out IMount mount);
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Script/Ambients/AmbientContext.cs]: >DominoScript [](start = 51, length = 12)

Please don't use DominoScript. Either use "DScript" or "Script"

File: Public/Src/FrontEnd/Script/Ambients/AmbientContext.cs [hunk 3]
```diff
@@ -223,13 +223,13 @@ private EvaluationResult GetMount(Context context, ModuleLiteral env, Evaluation
...
 
-        private static EvaluationResult HasMount(Context context, ModuleLiteral env, EvaluationStackFrame args)
+        private EvaluationResult HasMount(Context context, ModuleLiteral env, EvaluationStackFrame args)
         {
             var engine = context.FrontEndHost.Engine;
             var name = Convert.ToString(args[0].Value, CultureInfo.InvariantCulture);
             GetProvenance(context, out AbsolutePath path, out LineInfo lineInfo);
 
-            var result = engine.TryGetMount(name, "DominoScript", new BuildXL.Utilities.ModuleId(0), out IMount mount);
+            var result = engine.TryGetMount(name, "DominoScript", AmbientTypes.ScriptModuleId, out IMount mount);
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Script/Ambients/AmbientContext.cs]: >DominoScript [](start = 51, length = 12)

no "DominoScript" please

File: Public/Src/FrontEnd/Script/Ambients/AmbientGlobal.cs [hunk 4]
```diff
@@ -46,13 +46,15 @@ private List<NamespaceFunctionDefinition> GetGlobalFunctionDefinitions()
...
-                Function(Constants.Names.GlobFunction, this.Glob, GlobSignature),
+                Function(Constants.Names.GlobFunction, Glob, GlobSignature),
                 Function(Constants.Names.GlobRFunction, globRecursivelyBinding),
                 Function(Constants.Names.GlobRecursivelyFunction, globRecursivelyBinding),
-                Function(Constants.Names.GlobFoldersFunction, this.GlobFolders, GlobFoldersSignature),
+                Function(Constants.Names.GlobFoldersFunction, GlobFolders, GlobFoldersSignature),
                 Function("addIf", AddIf, AddIfSignature),
                 Function("addIfLazy", AddIfLazy, AddIfLazySignature),
                 Function("sign", Sign, SignSignature),
+                Function("unsafe_outputFile", UnsafeOutputFile, UnsafeOutputFileSignature),
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Script/Ambients/AmbientGlobal.cs]: Please place within new namespace Unsafe

File: Public/Src/FrontEnd/Script/Ambients/AmbientGlobal.cs [hunk 5]
```diff
@@ -94,6 +96,20 @@ private EvaluationResult AddIfLazy(Context context, ModuleLiteral env, Evaluatio
             return result;
         }
 
+        private EvaluationResult UnsafeOutputFile(Context context, ModuleLiteral env, EvaluationStackFrame args)
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Script/Ambients/AmbientGlobal.cs]: These features should NOT be exposed by default.
Please update this to check if this is enabled in the configuration. (no need to put in commandline, but at least it should be in config.dsc)

```
config({
   frontend: {
        unsafeFileApisAllowed: true
   }
});
```

Comment (dannyvv) [reply Public/Src/FrontEnd/Script/Ambients/AmbientGlobal.cs]: These methods should trhow

---
In reply to: [313650448](https://github.com/microsoft/BuildXL/pull/730#discussion_r313650448) [](ancestors = 313650448)

**Comment by dannyvv** [comment]
>     public readonly struct ModuleId : IEquatable<ModuleId>

I thought we agreed you'd add Version here as well to keep track of the version, but just not use it in the equals or hashcode yet.

---
Refers to: Public/Src/Utilities/Utilities/ModuleId.cs:16 in bd317fc. [](commit_id = bd317fc6e73752a76ed801b869ae1a697a325b6e, deletion_comment = False)

File: Public/Src/Utilities/Configuration/IFrontEndConfiguration.cs [hunk 6]
```diff
@@ -299,6 +299,11 @@ public interface IFrontEndConfiguration
         /// </remarks>
         IReadOnlyList<string> GlobalUnsafePassthroughEnvironmentVariables { get; }
 
+        /// <summary>
+        /// Wheather or not to release workspace before evaluation.
+        /// </summary>
+        bool ReleaseWorkspaceBeforeEvaluation { get; }
+
         /// <summary>
```
Comment (dannyvv) [inline Public/Src/Utilities/Configuration/IFrontEndConfiguration.cs]: Please add a flag here for UnsafeFileCreationAllowed

File: Public/Src/IDE/BuildXL.Ide.dsc [hunk 7]
```diff
@@ -8,12 +8,12 @@ export declare const qualifier: BuildXLSdk.DefaultQualifier;
...
-    targetFramework: "netcoreapp3.0";
-    targetRuntime: "win-x64" | "osx-x64";
-}
-
-namespace LanguageService {
-    export declare const qualifier : VsCodeExtensionQualifier;
-}
\ No newline at end of file
+//export interface VsCodeExtensionQualifier extends Qualifier {
+//    configuration: "debug" | "release";
```
Comment (dannyvv) [inline Public/Src/IDE/BuildXL.Ide.dsc]: remove or not check in...

File: Public/Src/FrontEnd/Ninja/NinjaWorkspaceResolver.cs [hunk 8]
```diff
@@ -268,7 +268,7 @@ private async Task<Possible<NinjaGraphWithModuleDefinition>> TryComputeBuildGrap
             var result = maybeGraph.Result;
             var specFileConfig = SpecFile.ChangeExtension(m_context.PathTable, PathAtom.Create(m_context.StringTable, ".ninja.dsc"));   // It needs to be a .dsc for the parsing to work
 
-            var moduleDescriptor = ModuleDescriptor.CreateWithUniqueId(m_resolverSettings.ModuleName, this);   
+            var moduleDescriptor = ModuleDescriptor.CreateWithUniqueId(m_context.StringTable, m_resolverSettings.ModuleName, this);
```
Comment (dannyvv) [inline Public/Src/FrontEnd/Ninja/NinjaWorkspaceResolver.cs]: >    [](start = 131, length = 3)

nit: extra whitespace

File: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+using BuildXL.Utilities.Instrumentation.Common;
+using JetBrains.Annotations;
+using static BuildXL.Scheduler.Graph.PipGraph;
+
+namespace BuildXL.Scheduler.Graph
+{
+    /// <summary>
+    /// Class for building graph fragments.
+    /// </summary>
+    public class GraphFragmentBuilder : IPipGraph
```
Comment (dannyvv) [inline Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilder.cs]: Could we mvoe this to BuildXL.Pips assembly?

This should be general use and not have to tie fully into the schedule which almost pulls in the entire engine

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 10]
```diff
@@ -20,15 +22,24 @@ namespace BuildXL.Scheduler.Graph
     /// </summary>
     public class PipGraphFragmentManager : IPipGraphFragmentManager
```
Comment (dannyvv) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: Ditto, can fragment management move to BuildXl.Pips assemlby?

**Comment by ljcollins25** [comment]
>     public readonly struct ModuleId : IEquatable<ModuleId>

I think Version is not needed since a identity uniquely identifies a module (a module may be a versioned entity but two different versioned entities with the same name, should have different module identities).

---
In reply to: [521044451](https://github.com/microsoft/BuildXL/pull/730#issuecomment-521044451) [](ancestors = 521044451)

---
Refers to: Public/Src/Utilities/Utilities/ModuleId.cs:16 in bd317fc. [](commit_id = bd317fc6e73752a76ed801b869ae1a697a325b6e, deletion_comment = False)

File: Public/Sdk/Public/Prelude/Prelude.IO.dsc (hunk 1 above)
Comment (narasamdya) [reply Public/Sdk/Public/Prelude/Prelude.IO.dsc]: I created `Prelude.Unsafe.dsc`

---
In reply to: [313613352](https://github.com/microsoft/BuildXL/pull/730#discussion_r313613352) [](ancestors = 313613352)

**Comment by narasamdya** [comment]
Discussed the reason offline.

---
In reply to: [521010296](https://github.com/microsoft/BuildXL/pull/730#issuecomment-521010296) [](ancestors = 521010296)

File: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilder.cs (hunk 9 above)
Comment (narasamdya) [reply Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilder.cs]: It has the functionality as a graph builder. In the next stage, we're going to infer some dependencies to make graph deserialization parallel.

---
In reply to: [313653006](https://github.com/microsoft/BuildXL/pull/730#discussion_r313653006) [](ancestors = 313653006)

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs (hunk 10 above)
Comment (narasamdya) [reply Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: This class adds deserialized pips to pip graph. So, I think it's natural to place it in the PipGraph assembly.

---
In reply to: [313653201](https://github.com/microsoft/BuildXL/pull/730#discussion_r313653201) [](ancestors = 313653201)

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 11]
```diff
@@ -20,15 +22,24 @@ namespace BuildXL.Scheduler.Graph
...
 
-        private PipGraphFragmentContext m_fragmentContext;
+        private readonly LoggingContext m_loggingContext;
 
-        private LoggingContext m_loggingContext;
+        private readonly ConcurrentBigMap<ModuleId, Lazy<bool>> m_modulePipUnify = new ConcurrentBigMap<ModuleId, Lazy<bool>>();
+
+        private readonly ConcurrentBigMap<FileArtifact, Lazy<bool>> m_specFilePipUnify = new ConcurrentBigMap<FileArtifact, Lazy<bool>>();
+
+        private readonly ConcurrentBigMap<(FullSymbol, QualifierId), Lazy<bool>> m_valuePipUnify = new ConcurrentBigMap<(FullSymbol, QualifierId), Lazy<bool>>();
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >m_valuePipUnify [](start = 81, length = 15)

I think containing spec file needs to be in the key as well.

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs [hunk 12]
```diff
@@ -20,15 +22,24 @@ namespace BuildXL.Scheduler.Graph
...
+
+        private readonly ConcurrentBigMap<long, Lazy<bool>> m_pipUnify = new ConcurrentBigMap<long, Lazy<bool>>();
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: >m_pipUnify [](start = 60, length = 10)

Maybe add a comment that mentions the key here is the semistable hash.

File: Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs (hunk 10 above)
Comment (dannyvv) [reply Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs]: note we don't have a pipgraph assembly, this is BuildXL.Schedule.dll :)

---
In reply to: [313718382](https://github.com/microsoft/BuildXL/pull/730#discussion_r313718382) [](ancestors = 313718382,313653201)

**Pull request merged**